### PR TITLE
more tests and lints for feature detect.js

### PR DESCRIPTION
**Pull request by romainmenke**
resolves : https://github.com/Financial-Times/polyfill-library/issues/520

- add a [linter rule plugin](https://github.com/romainmenke/eslint-plugin-uncalled-iife) to detect uncalled `IIFE` in `detect.js`
- ~~verify that calling `detect.js` as a function throws~~
- fix some `detect.js` files
- fix `DocumentFragment` polyfill

----------

### eslint-plugin-uncalled-iife

I would have preferred to include a custom eslint rule in the `polyfill-library` repo but `eslint` does not support this.
The only way was to create a separate package containing the rule and publish it on npm.
This package is really small and I do not mind maintaining it when eslint introduces breaking changes.

There are cases that can't be caught this way. (named functions that aren't called)

Screenshot 2020-10-16 at 11 15 30

### DocumentFragment

Testing for uncalled `IIFE`'s exposed some issues with the `DocumentFragment ` polyfill.
I do not see how the previous code could have worked, but I might be overlooking something.
If you prefer, I can split this into two PR's.

### test for falsy

I tried adding a test that verifies that the code in `detect.js` returns something falsy in browsers that need the polyfill when the polyfill isn't loaded. In theory this would expose false positives/negatives. In practice this doesn't work :) Too many features would pass a `detect.js` but still need the polyfill. This then required adding a lot of exceptions which makes the test less useful.

### CODE REVIEW

**Comment by JakeChampion** [comment]
@romainmenke This is superb!

File: test/polyfills/server.js [hunk 1]
```diff
@@ -163,6 +163,12 @@ async function testablePolyfills(isIE8, ua) {
           }).call(window));
         });
 
+        it('throws when called as a function', function() {
+          proclaim["throws"](function () {
+            (${config.detectSource})();
+          });
+        });
+
```
Comment (JakeChampion) [inline test/polyfills/server.js:171]: Why is this needed? Does it cover things the eslint rule does not?

Comment (romainmenke) [reply test/polyfills/server.js:171]: This catches `function foo(){ return true }` (vs. `function foo(){ return true } && foo()`)

I am not sure if we need or want this test.
These are edge cases and PR review might be a better way to intercept possible issues?

Comment (JakeChampion) [reply test/polyfills/server.js:171]: Yeah, I think we should remove this test and rely on PR review 👍

Comment (romainmenke) [reply test/polyfills/server.js]: Cool! done

**Comment by romainmenke** [comment]
@JakeChampion seems a branch was created here : https://github.com/Financial-Times/polyfill-library/tree/more-tests-and-lints-for-feature-detect--passionate-australian-mist-84fb32e692

A bit weird, but can be removed.

**Comment by JakeChampion** [comment]
@romainmenke yep, I need to push the same commits to a branch in the origin (Financial-Times/polyfill-library) in order to have CI to run with the BrowserStack credentials

**Comment by romainmenke** [comment]
Good to know :)
